File: leetcode/leetcode015.cpp
//O(n^2)复杂度，在一个数组中，找到所有满足三个数和为0的组合。
#include <iostream>
#include <vector>
using namespace std;
class Solution
{
public:
	vector<vector<int>> threeSum(vector<int>& nums){
		vector<vector<int>> result;
		int i,l,r,n=nums.size();
		int s;
		sort(nums.begin(),nums.end());
		for(i=0;i<n-2;i++){
			if (i>0 && nums[i]==nums[i-1])
			{
				continue;
			}
			l=i+1;
			r=n-1;
			while(r<l){
			s=nums[i]+nums[l]+nums[r];
			if(s<0) l++;
			else if(s>0) r--;
			else{
				vector<int> tmp;
				tmp.push_back(nums[i]);
				tmp.push_back(nums[l]);
				tmp.push_back(nums[r]);
				result.push_back(tmp);
				while(l<r && nums[l]==nums[l+1]) l++;
				while(l<r && nums[r]==nums[r-1]) r--;
				l++;r--;
			}
		}
	}
		return result;
	}
};

File: leetcode/leetcode015.py
#!/usr/bin/python
#coding=utf-8
#3Sum
#给定一组数组，找到该整数数组中满足和为0的三个数字组合。
class Solution(object):
	def threeSum(self,nums):
		result=[]
		num1=sorted(nums)
		n=len(num1)
		for x in range(0,n):
			if x==0 or (x>0 and num1[x]!=num1[x-1]):
				target=-num1[x]
				l,r=x+1,n-1
				while l<r:
					if num1[l]+num1[r]==target:
						result.append([num1[x],num1[l],num1[r]])
						while l<r and num1[l]==num1[l+1]: l+=1
						while l<r and num1[r]==num1[r-1]: r-=1
						l+=1
						r-=1
					elif num1[l]+num1[r]<target:
						l+=1
					else:
						r-=1
		return result
a=Solution()
S = [-1, 0, 1, 2, -1, -4]
b=a.threeSum(S)
print b